### PR TITLE
MGMT-15814: Ensure that preparation timeout gives some clues

### PR DESCRIPTION
**Pull request by paul-maidment**
When preparation times out, we would like to provide some clues as to why the timeout occurred. By obtaining the validations of the host that are not successful, we can provide an indication of this.

## List all the issues related to this PR

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [] None

## How was this code tested?

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/cluster/transition.go [hunk 1]
```diff
@@ -524,10 +525,50 @@ func (th *transitionHandler) PostPreparingTimedOut(sw stateswitch.StateSwitch, a
 		updatedCluster *common.Cluster
 	)
 
+	// Capture some information about hosts states, we should be able to determine more about the cause of the timeout.
+	// This can occur because the host...
+	//
+	// 1: Becomes disconnected - the "connected" validation of the host will be failing
```
Comment (filanov) [inline internal/cluster/transition.go:531]: disconnected is handled by another transition https://github.com/openshift/assisted-service/blob/2c70724e4b5f79c472847a0d78dcd77f0f2a0197/internal/host/statemachine.go#L696C1-L710

File: internal/cluster/transition.go [hunk 2]
```diff
@@ -524,10 +525,50 @@ func (th *transitionHandler) PostPreparingTimedOut(sw stateswitch.StateSwitch, a
...
+			// Deserialize validations info
+			var hostValidations ValidationsStatus
+			err = json.Unmarshal([]byte(clusterHost.ValidationsInfo), &hostValidations)
+			if err != nil {
+				issues = append(issues, fmt.Sprintf("unable to deserialize validations info for host %s due to error %s", clusterHost.ID, err.Error()))
+			}
+			for _, vRes := range hostValidations {
+				for _, v := range vRes {
+					// Any unsuccesful validation should be noted as these are likely to have caused issues
+					if v.Status != ValidationSuccess {
```
Comment (filanov) [inline internal/cluster/transition.go:551]: I'm not sure if this part is correct, in the timeout that i saw the result of the validation was success but didn't run.
If i remember correctly the validation between the condition and the validation info is different, the validation marked as success with the info that state that it didn't run yet, but the condition is failing.

**Comment by filanov** [comment]
how was it tested in the e2e flow?

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -1291,6 +1291,12 @@ func (b *bareMetalInventory) InstallClusterInternal(ctx context.Context, params
 
 	// prepare cluster and hosts for installation
 	err = b.db.Transaction(func(tx *gorm.DB) error {
+		if err = b.db.Model(&common.Cluster{}).Where("id = ?", params.ClusterID).Updates(&common.Cluster{
```
Comment (filanov) [inline internal/bminventory/inventory.go:1294]: why are you using `b.db`? if you are inside a transaction you should use tx

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -1354,6 +1356,14 @@ func (b *bareMetalInventory) InstallClusterInternal(ctx context.Context, params
 				return
 			}
 		}
+
+		// Mark the async cluster preparation as finished
+		err = b.db.Transaction(func(tx *gorm.DB) error {
```
Comment (filanov) [inline internal/bminventory/inventory.go:1361]: why are you opening a transaction for a single operation?

File: internal/cluster/transition.go [hunk 5]
```diff
@@ -542,7 +542,12 @@ func (th *transitionHandler) PostPreparingTimedOut(sw stateswitch.StateSwitch, a
 		return err
 	}
 
-	eventgen.SendInstallationPreparingTimedOutEvent(params.ctx, params.eventHandler, *cluster.ID)
+	causedBy := "Preparation of configuration for installation of the cluster was started but did not complete. In the context of a timeout this is most likely to indicate that there was a problem with an assisted-service pod, please retry the installation."
```
Comment (filanov) [inline internal/cluster/transition.go:545]: suggesting to consult with UX to make the message as clear as possible @nirfarkas

Comment (nirfarkas) [reply internal/cluster/transition.go:545]: The message should clearly state:
What just went wrong?
Why did this happen?
How to solve it?

Comment (paul-maidment) [reply internal/cluster/transition.go:545]: In terms of why, we can't say for certain that the assisted pod crashed, only that for whatever reason the async goroutine to perform preparation did not complete.

I think the phrases we have suggested encapsulate this, or would you suggest alternative wording?

Comment (paul-maidment) [reply internal/cluster/transition.go]: I'm going with 

"An unexpected error occurred in the assisted installer, please retry installation and contact support if the issue persists"

and 

"One or more cluster hosts were not prepared successfully, please check host events for more information"

File: internal/cluster/transition.go [hunk 6]
```diff
@@ -509,6 +509,17 @@ func (th *transitionHandler) IsPreparingTimedOut(sw stateswitch.StateSwitch, arg
 	return false, nil
 }
 
+func (th *transitionHandler) areAllHostsPreparedForInstallation(cluster *common.Cluster) bool {
+	preparedHostCount := 0
+	for _, clusterHost := range cluster.Hosts {
+		if *clusterHost.Status != models.HostStatusPreparingSuccessful {
+			break
```
Comment (ori-amizur) [inline internal/cluster/transition.go:516]: Here you can just return false

File: internal/cluster/transition.go [hunk 7]
```diff
@@ -509,6 +509,17 @@ func (th *transitionHandler) IsPreparingTimedOut(sw stateswitch.StateSwitch, arg
...
+		}
+		preparedHostCount++
+	}
+	return preparedHostCount == len(cluster.Hosts)
```
Comment (ori-amizur) [inline internal/cluster/transition.go:520]: Here you can return true

File: internal/cluster/transition.go [hunk 8]
```diff
@@ -509,6 +509,17 @@ func (th *transitionHandler) IsPreparingTimedOut(sw stateswitch.StateSwitch, arg
 	return false, nil
 }
 
+func (th *transitionHandler) areAllHostsPreparedForInstallation(cluster *common.Cluster) bool {
+	preparedHostCount := 0
```
Comment (ori-amizur) [inline internal/cluster/transition.go:513]: Not needed (the variable)

File: internal/cluster/transition.go [hunk 9]
```diff
@@ -509,6 +509,17 @@ func (th *transitionHandler) IsPreparingTimedOut(sw stateswitch.StateSwitch, arg
...
+	for _, clusterHost := range cluster.Hosts {
+		if *clusterHost.Status != models.HostStatusPreparingSuccessful {
```
Comment (ori-amizur) [inline internal/cluster/transition.go:515]: use `swag.StringValue` (safety)

**Comment by paul-maidment** [comment]
/hold
Need to set vars for LastInstallationPreparation status

File: internal/cluster/transition.go [hunk 10]
```diff
@@ -542,7 +551,23 @@ func (th *transitionHandler) PostPreparingTimedOut(sw stateswitch.StateSwitch, a
 		return err
 	}
 
-	eventgen.SendInstallationPreparingTimedOutEvent(params.ctx, params.eventHandler, *cluster.ID)
+	allHostsPrepared := th.areAllHostsPreparedForInstallation(cluster)
```
Comment (ori-amizur) [inline internal/cluster/transition.go:554]: I thought that hosts are handling their timeout.  Why are we handling this here ?

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 11]
```diff
@@ -2009,6 +2010,27 @@ func clusterCompleted(clusterInstall *hiveext.AgentClusterInstall, clusterDeploy
 	})
 }
 
+func lastInstallationPreparationFailed(clusterInstall *hiveext.AgentClusterInstall, installation_preparation_completion_status string, installation_preparation_completion_status_reason string) {
+	var condStatus corev1.ConditionStatus = corev1.ConditionFalse
```
Comment (ori-amizur) [inline internal/controller/controllers/clusterdeployments_controller.go:2014]: Is this function called in all cases? Including when installation is still in progress, and if it succeeds?
If yes, please rename to indicate what it is actually doing

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 12]
```diff
@@ -2009,6 +2010,27 @@ func clusterCompleted(clusterInstall *hiveext.AgentClusterInstall, clusterDeploy
...
+	var reason string = hiveext.ClusterLastInstallationPreparationFailedOKReason
+	var msg string = ""
+
+	switch installation_preparation_completion_status {
```
Comment (ori-amizur) [inline internal/controller/controllers/clusterdeployments_controller.go:2018]: Please let the switch-case handle all possible cases.  It is not clear from the code.

**Pull request merged**